compiler/rustc_lint/src/internal.rs

```diff
@@ -122,7 +125,121 @@ impl LateLintPass<'_> for QueryStability {
                 );
             }
         }
+        check_into_iter_stability(cx, expr);
+    }
+}
+
+fn check_into_iter_stability<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+    let Some(into_iterator_def_id) = cx.tcx.get_diagnostic_item(sym::IntoIterator) else {
+        return;
+    };
+    if expr.span.from_expansion() {
+        return;
+    };
+    // Is `expr` a function or method call?
+    let Some((callee_def_id, generic_args, recv, args)) =
+        get_callee_generic_args_and_args(cx, expr)
+    else {
+        return;
+    };
+    let fn_sig = cx.tcx.fn_sig(callee_def_id).instantiate_identity().skip_binder();
+    for (arg_index, &input) in fn_sig.inputs().iter().enumerate() {
```

Comment:
In general, this implementations seems overly complex to me for what it's checking. You're basically implementing an very limited ad-hoc trait solver:

1. going through all predicates in the param env,
2. eliminating uneligible candidates,
3. going through all impls,
4. eliminating impls where the self ty doesn't match (based on `Ty`'s `==` which is ptr equality instead of type relating like `ObligationCtxt::eq(…)`)

and all of that in ~O(n^3).

I wonder we could utilize some existing trait solving methods :thinking: I need to think about it. Sorry, it's not very actionable I g2g

Reply:
You were right: the lint has a much simpler implementation using `Instance::try_resolve`. I have verified that the new implementation necessitates [all of the changes](https://github.com/rust-lang/rust/pull/139345/commits/1a2c5a35b9d4a8258aa7f5f9996da842f93a7d20) the previous one did.